### PR TITLE
:sparkles: Add guild sheduled event endpoints

### PR DESCRIPTION
**Pull request by beastmatser**
### Changes

-   `adds`: The endpoints for the guild scheduled events

 ### Check off the following

-   [x] I have tested my changes with the current requirements
-   [x] My Code follows the pep8 code style.

### CODE REVIEW

File: pincer/objects/guild/guild.py [hunk 1]
```diff
@@ -1962,6 +1964,123 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        scheduled_end_time: Optional[datetime] = None,
+        entity_metadata: Optional[str] = None,
+        channel_id: Optional[Snowflake] = None,
+        description: Optional[str] = None,
+        reason: Optional[str] = None,
+    ) -> ScheduledEvent:
+        if scheduled_start_time.timestamp() < datetime.now().timestamp():
+            raise ValueError("An event cannot be created in the past")
+
+        if scheduled_end_time is not None and scheduled_end_time.timestamp() < datetime.now().timestamp():
```
Comment (Sigmanificient) [inline pincer/objects/guild/guild.py:1945]: The line length seems to be quite long, did you forgot to run black?

File: pincer/objects/guild/guild.py [hunk 2]
```diff
@@ -1962,6 +1964,123 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        channel_id: Optional[Snowflake] = None,
+        description: Optional[str] = None,
+        status: Optional[int] = None,
+        reason: Optional[str] = None,
+    ) -> ScheduledEvent:
+        if scheduled_start_time is not None and scheduled_start_time.timestamp() < datetime.now().timestamp():
+            raise ValueError("An event cannot be created in the past")
+
+        if scheduled_end_time is not None and scheduled_end_time.timestamp() < datetime.now().timestamp():
+            raise ValueError("An event cannot start before it ends")
```
Comment (Sigmanificient) [inline pincer/objects/guild/guild.py:1990]:
```suggestion
        if scheduled_start_time is not None:
            timestamp = datetime.now().timestamp()
            if scheduled_start_time.timestamp() < timestamp:
                raise ValueError("An event cannot be created in the past")

            if scheduled_end_time.timestamp() < timestamp:
                raise ValueError("An event cannot start before it ends")
```
Avoiding calculating twice the same thing

File: pincer/objects/guild/guild.py [hunk 3]
```diff
@@ -1962,6 +1964,123 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        data = await self._http.patch(
+            f"guilds/{self.id}/scheduled-events/{_id}",
+            data=kwargs,
+            headers={"X-Audit-Log-Reason": reason},
+            )
+        return ScheduledEvent.from_dict(
+            construct_client_dict(self._client, data)
+        )
+
+    async def delete_scheduled_event(self, _id: Snowflake):
```
Comment (Sigmanificient) [inline pincer/objects/guild/guild.py:2013]:
```suggestion
    async def delete_scheduled_event(self, _id: Union[int, Snowflake]):
```

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: pincer/objects/guild/guild.py [hunk 4]
```diff
@@ -1916,6 +1919,133 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+            },
+            headers={"X-Audit-Log-Reason": reason},
+        )
+        return ScheduledEvent.from_dict(data)
+
+    async def get_scheduled_event(
+        self, _id: Snowflake, with_user_count: bool = False
+    ) -> ScheduledEvent:
+        data = await self._http.get(
+            f"guilds/{self.id}/scheduled-events/{_id}?{with_user_count=!s}"
```
Comment (Lunarmagpie) [inline pincer/objects/guild/guild.py:1974]: You should use the params argument here

File: pincer/objects/guild/guild.py [hunk 5]
```diff
@@ -1916,6 +1919,133 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        )
+
+        data = await self._http.patch(
+            f"guilds/{self.id}/scheduled-events/{_id}",
+            data=kwargs,
+            headers={"X-Audit-Log-Reason": reason},
+        )
+        return ScheduledEvent.from_dict(self._client, data)
+
+    async def delete_scheduled_event(self, _id: Union[int, Snowflake]):
```
Comment (Lunarmagpie) [inline pincer/objects/guild/guild.py:2026]: I don't think you need to specify int if snowflake is there. No sure though.

File: pincer/objects/guild/guild.py [hunk 6]
```diff
@@ -1916,6 +1919,133 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+    async def get_guild_scheduled_event_users(
+        self,
+        _id: Snowflake,
+        limit: int = 100,
+        with_member: bool = False,
+        before: Optional[Snowflake] = None,
+        after: Optional[Snowflake] = None,
+    ) -> AsyncGenerator[GuildScheduledEventUser, None]:
+        url = f"guilds/{self.id}/scheduled-events/{_id}/users?{limit=!s}&{with_member=!s}"
+
```
Comment (Lunarmagpie) [inline pincer/objects/guild/guild.py:2038]: The `params` argument for http.get should be used here to

File: pincer/objects/guild/scheduled_events.py [hunk 7]
```diff
@@ -112,3 +113,22 @@ class ScheduledEvent(APIObject):
...
+    """
+    Represents a user who has joined a scheduled event.
+
+    Attributes
+    ----------
+    guild_scheduled_event_id: :class:`int`
+        the scheduled event id which the user subscribed to
+    user : :class:`~pincer.objects.user.user.User`
+        user which subscribed to an event
+    member : :class:`~pincer.objects.guild.member.GuildMember`
```
Comment (Lunarmagpie) [inline pincer/objects/guild/scheduled_events.py:129]: This should be APINullable

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]

File: pincer/objects/guild/guild.py (hunk 5 above)
Comment (beastmatser) [reply pincer/objects/guild/guild.py:2026]: A Snowflake is an int but an int is not necessarily a Snowflake, so yes?

Comment (Sigmanificient) [reply pincer/objects/guild/guild.py:2029]: # Typing with Snowflake
```py
def bar(x: Snowflake):
    print(x + 69)


bar(69)
bar(Snowflake(69))
```
> ```
> def.py:5: error: Argument 1 to "bar" has incompatible type "int"; expected "Snowflake"  [arg-type]
> Found 1 error in 1 file (checked 1 source file)
> ```


# Typing with Int
```py
def baz(x: int):
    print(x + 42)


baz(42)
baz(Snowflake(42))
```
> ```
Success: no issues found in 1 source file
>```

Comment (Lunarmagpie) [reply pincer/objects/guild/guild.py:2029]: Ig always type with int

**Review by Lunarmagpie — APPROVED** [review]

File: pincer/objects/guild/guild.py [hunk 8]
```diff
@@ -1916,6 +1919,266 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+            The newly created scheduled event.
+        """
+        if scheduled_start_time.timestamp() < datetime.now().timestamp():
+            raise ValueError("An event cannot be created in the past")
+
+        if (
+            scheduled_end_time is not None
+            and scheduled_end_time.timestamp() < datetime.now().timestamp()
+        ):
+            raise ValueError("An event cannot start before it ends")
```
Comment (trag1c) [inline pincer/objects/guild/guild.py:1992]:
```suggestion
        if scheduled_start_time < datetime.now():
            raise ValueError("An event cannot be created in the past")

        if (
            scheduled_end_time is not None
            and scheduled_end_time < datetime.now()
        ):
            raise ValueError("An event cannot start before it ends")
```

File: pincer/objects/guild/guild.py [hunk 9]
```diff
@@ -1916,6 +1919,266 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        data = await self._http.post(
+            f"guilds/{self.id}/scheduled-events",
+            data={
+                "name": name,
+                "privacy_level": privacy_level,
+                "entity_type": entity_type,
+                "scheduled_start_time": scheduled_start_time.isoformat(),
+                "scheduled_end_time": scheduled_end_time.isoformat()
+                if scheduled_end_time is not None
+                else None,
```
Comment (trag1c) [inline pincer/objects/guild/guild.py:2003]: :thinking: :thinking: :thinking: 
```suggestion
                "scheduled_end_time": getattr(
                    scheduled_end_time, "isoformat", type(None)
                )(),
```

File: pincer/objects/guild/guild.py [hunk 10]
```diff
@@ -1916,6 +1919,266 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        """
+        if scheduled_start_time:
+            timestamp = scheduled_start_time.timestamp()
+            if timestamp < datetime.now().timestamp():
+                raise ValueError("An event cannot be created in the past")
+
+            if (
+                scheduled_end_time
+                and timestamp > scheduled_end_time.timestamp()
+            ):
```
Comment (trag1c) [inline pincer/objects/guild/guild.py:2097]:
```suggestion
            if scheduled_start_time < datetime.now():
                raise ValueError("An event cannot be created in the past")

            if (
                scheduled_end_time
                and scheduled_start_time > scheduled_end_time
            ):
```

File: pincer/objects/guild/guild.py [hunk 11]
```diff
@@ -1916,6 +1919,266 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+            {
+                "name": name,
+                "privacy_level": privacy_level,
+                "entity_type": entity_type,
+                "scheduled_start_time": scheduled_start_time.isoformat()
+                if scheduled_start_time is not None
+                else None,
+                "scheduled_end_time": scheduled_end_time.isoformat()
+                if scheduled_end_time is not None
+                else None,
```
Comment (trag1c) [inline pincer/objects/guild/guild.py:2110]: Maybe same here?
```suggestion
                "scheduled_start_time": getattr(
                    scheduled_start_time, "isoformat", type(None)
                )(),
                "scheduled_end_time": getattr(
                    scheduled_end_time, "isoformat", type(None)
                )(),
```

File: pincer/objects/guild/guild.py [hunk 12]
```diff
@@ -1916,6 +1919,266 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        }
+        if before is not None:
+            params.update({"before": before})
+        if after is not None:
+            params.update({"after": after})
+
+        data = await self._http.get(
+            f"guilds/{self.id}/scheduled-events/{_id}/users?",
+            params=params,
+        )
```
Comment (trag1c) [inline pincer/objects/guild/guild.py:2177]:
```suggestion
        params = {
            "limit": limit,
            "with_member": with_member,
            "before": before,
            "after": after
        }

        data = await self._http.get(
            f"guilds/{self.id}/scheduled-events/{_id}/users",
            params=remove_none(params),
        )
```

**Review by trag1c — CHANGES_REQUESTED** [review]

File: pincer/objects/guild/guild.py (hunk 11 above)
Comment (beastmatser) [reply pincer/objects/guild/guild.py:2110]: I don't think I'll change this, I find that this change is not very readable

Comment (trag1c) [reply pincer/objects/guild/guild.py:2110]: Alrighty, resolve my comments then

File: pincer/objects/guild/guild.py [hunk 13]
```diff
@@ -1916,6 +1919,264 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        name: str,
+        privacy_level: int,
+        entity_type: int,
+        scheduled_start_time: datetime,
+        scheduled_end_time: Optional[datetime] = None,
+        entity_metadata: Optional[str] = None,
+        channel_id: Optional[int] = None,
+        description: Optional[str] = None,
+        reason: Optional[str] = None,
+    ) -> ScheduledEvent:
```
Comment (trag1c) [inline pincer/objects/guild/guild.py:1956]: Missing `Raises` section in the docs

File: pincer/objects/guild/guild.py [hunk 14]
```diff
@@ -1916,6 +1919,264 @@ async def get_webhooks(self) -> AsyncGenerator[Webhook, None]:
...
+        await self._http.delete(f"guilds/{self.id}/scheduled-events/{_id}")
+
+    async def get_guild_scheduled_event_users(
+        self,
+        _id: int,
+        limit: int = 100,
+        with_member: bool = False,
+        before: Optional[int] = None,
+        after: Optional[int] = None,
+    ) -> AsyncGenerator[GuildScheduledEventUser, None]:
```
Comment (trag1c) [inline pincer/objects/guild/guild.py:2143]: Same as above I guess?
```suggestion
    ) -> AsyncIterator[GuildScheduledEventUser]:
```

**Review by trag1c — APPROVED** [review]

**Review by Sigmanificient — APPROVED** [review]

**Pull request merged**